Source/ProgressionSystemRuntime/Public/Data/PSSaveGameData.h

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Valerii Rotermel & Yevhenii Selivanov
+
+#pragma once
+
+#include "Data/PSTypes.h"
+#include "Templates/Tuple.h"
+#include "GameFramework/SaveGame.h"
+#include "PSSaveGameData.generated.h"
+
+
+/**
+ * Defines the standard process for the saving slots names and index 
+ */
+UCLASS()
+class PROGRESSIONSYSTEMRUNTIME_API UPSSaveGameData : public USaveGame
+{
+	GENERATED_BODY()
+
+public:
+	/** Returns the name of the save slot. */
+	UFUNCTION(BlueprintPure, Category = "C++")
+	static const FString& GetSaveSlotName();
+
+	/** Returns the Slot Index of the save slot. */
+	UFUNCTION(BlueprintPure, Category = "C++")
+	static int32 GetSaveSlotIndex() { return 0; }
+
+	/** Returns the ProgressionRow by Index */
+	UFUNCTION(BlueprintPure, Category = "C++")
+	const FPSRowData& GetSavedProgressionRowByIndex(int32 Index) const;
+
+	/** Set element by index */
+	UFUNCTION(BlueprintCallable, Category= "C++")
+	void SetCurrentProgressionRowByIndex(int32 InIndex);
+
+	/** Update the ProgressionRows map */
+	UFUNCTION(BlueprintCallable, Category = "C++")
+	void SetProgressionMap(const TMap<FName, FPSRowData> ProgressionRows);
+
+	/** Get current progression row */
+	UFUNCTION(BlueprintPure, Category = "C++")
+	const FPSRowData& GetCurrentRow() const;
+
+	/** Get current progression row */
+	UFUNCTION(BlueprintPure, Category = "C++")
+	FORCEINLINE FName GetCurrentRowName() const { return CurrentRowName; }
+
+	/** Set the current level by player tag*/
+	void SetRowByTag(FPlayerTag PlayerTag);
+
+	/** Unlock level by Index, used only for the first level */
+	UFUNCTION(BlueprintCallable, Category = "C++")
+	void UnlockLevelByName(FName RowName);
+
+	/** Unlock level by Index, used only for the first level */
+	UFUNCTION(BlueprintCallable, Category = "C++")
+	void SavePoints(EEndGameState EndGameState);
+
+	/** Unlocks the next level*/
+	void NextLevelProgressionRowData();
+
+	/** Returns the endgame reward. */
+	UFUNCTION(BlueprintCallable, Category="C++")
+	int32 GetProgressionReward(EEndGameState EndGameState);
+
+protected:
+	/** The current Index of Saved Progression. */
+	UPROPERTY()
+	FName CurrentRowName = NAME_None;
+
+	/** The current Saved Progression of a player. */
+	UPROPERTY(VisibleInstanceOnly, BlueprintReadWrite, Category = "C++", meta = (BlueprintProtected, DisplayName = "Saved Progression Rows"))
+	TMap<FName, FPSRowData> SavedProgressionRows;
```

Comment:
Please call all protected properties as `Internal`, so it's easier to understand which properties are public and which are not.

Reply:
done